### PR TITLE
Add function for passing CMake flags as macros to applications

### PR DESCRIPTION
**Pull request by yourslab**
This was a common pattern in the CMake files of our demos and tests:
```
if(AWS_IOT_ENDPOINT)
    target_compile_definitions(
        ${stest_name} PRIVATE
            AWS_IOT_ENDPOINT="${AWS_IOT_ENDPOINT}"
    )
endif()
if(ROOT_CA_CERT_PATH)
    target_compile_definitions(
        ${stest_name} PRIVATE
            ROOT_CA_CERT_PATH="${ROOT_CA_CERT_PATH}"
    )
endif()
if(CLIENT_CERT_PATH)
    target_compile_definitions(
        ${stest_name} PRIVATE
        CLIENT_CERT_PATH="${CLIENT_CERT_PATH}"
    )
endif()
if(CLIENT_PRIVATE_KEY_PATH)
    target_compile_definitions(
        ${stest_name} PRIVATE
        CLIENT_PRIVATE_KEY_PATH="${CLIENT_PRIVATE_KEY_PATH}"
    )
endif()
```

This PR simplifies the above by creating a function that replicates this functionality and also checks whether certain required definitions are already defined in the header file or passed as a flag:
```
set_macro_definitions(TARGETS ${stest_name}
                      REQUIRED
                        "AWS_IOT_ENDPOINT"
                        "ROOT_CA_CERT_PATH"
                        "CLIENT_CERT_PATH"
                        "CLIENT_PRIVATE_KEY_PATH"
                        "CLIENT_IDENTIFIER"
                      OPTIONAL
                        "CLIENT_USERNAME"
                        "CLIENT_PASSWORD"
                        "OS_NAME"
                        "OS_VERSION"
                        "HARDWARE_PLATFORM_NAME")
```

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: tools/cmake/utility.cmake [hunk 1]
```diff
@@ -0,0 +1,68 @@
+function(set_macro_definitions)
```
Comment (yourslab) [inline tools/cmake/utility.cmake:1]: This is the meat of the PR

Comment (yourslab) [reply tools/cmake/utility.cmake:1]: Also, lemme know if you can think of a better function name for this.

File: tools/cmake/utility.cmake [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    foreach(application_target ${MACRO_DEFINITIONS_TARGETS})
+        if(DEFINED MACRO_DEFINITIONS_OPTIONAL)
+            foreach(optional_macro_definition ${MACRO_DEFINITIONS_OPTIONAL})
+                if(DEFINED ${optional_macro_definition})
+                    # Compile the application with the macro definition if it is defined.
+                    target_compile_definitions(
+                        ${application_target} PRIVATE
+                            ${optional_macro_definition}="${${optional_macro_definition}}"
+                    )
+                    list(APPEND DEFINED_MACROS "${required_macro_definition}")
```
Comment (sukhmanm) [inline tools/cmake/utility.cmake:25]: Should the optional macro (`optional_macro_definition`) be used here instead?

File: tools/cmake/utility.cmake (hunk 1 above)
Comment (sukhmanm) [reply tools/cmake/utility.cmake:1]: No objections to the name -- can't really think of a better way to describe this function.

File: tools/cmake/utility.cmake (hunk 2 above)
Comment (yourslab) [reply tools/cmake/utility.cmake]: Thanks for catching this

**Review by sukhmanm — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Pull request merged**